lib/active_merchant/billing/gateways/ixopay.rb

```diff
@@ -67,55 +54,193 @@ def scrub(transcript)
 
       private
 
-      def add_customer_data(post, options)
+      def headers(xml)
+        timestamp = Time.now.httpdate
+        signature = generate_signature('POST', xml, timestamp)
+
+        {
+          'Authorization' => "Gateway #{options[:api_key]}:#{signature}",
+          'Date' => timestamp,
+          'Content-Type' => 'text/xml; charset=utf-8'
+        }
+      end
+
+      def generate_signature(http_method, xml, timestamp)
+        content_type = 'text/xml; charset=utf-8'
+
+        message = http_method + "\n" + Digest::MD5.hexdigest(xml) + "\n" + content_type + "\n" \
+          + timestamp + "\n\n" + '/transaction'
+
+        digest = OpenSSL::Digest.new('sha512')
+        hmac = OpenSSL::HMAC.digest(digest, @secret, message)
+
+        Base64.encode64(hmac).split.join
+      end
+
+      def parse(action, xml)
+        parse_element({:action => action}, REXML::Document.new(xml))
```

Comment:
Consider whether is can be replaced with the default function: https://github.com/spreedly/internal-docs/blob/master/gateway_templates/example.rb#L194

Reply:
I think that this method should probably be replaced with a library method altogether. I'll leave it as a todo for the moment, for the sake of getting this PR merged into the integration branch ASAP since it's needed for other work on Ixopay.

Reply:
I don't, but can look into it after this branch is merged into the integration branch.